eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -104,44 +105,66 @@ import Prettyprinter (Pretty (..))
 -- =========================================================
 -- Translate Hashes, Credentials, Certificates etc.
 
-data TranslationError
-  = ByronInputInContext
-  | ByronOutputInContext
-  | TranslationLogicErrorInput
-  | RdmrPtrPointsToNothing RdmrPtr
-  | TranslationLogicErrorDoubleDatum
-  | LanguageNotSupported
-  | InlineDatumsNotSupported
-  | ReferenceScriptsNotSupported
-  | ReferenceInputsNotSupported
-  | TimeTranslationPastHorizon
+-- | A transaction output can be translated because it is a newly created output,
+-- or because it is the output which is connected to a transaction input being spent.
+data TxOutSource crypto
+  = TxOutFromInput !(TxIn crypto)
+  | TxOutFromOutput !TxIx
   deriving (Eq, Show, Generic, NoThunks)
 
-instance ToCBOR TranslationError where
-  toCBOR ByronInputInContext = encode $ Sum ByronInputInContext 0
-  toCBOR ByronOutputInContext = encode $ Sum ByronOutputInContext 1
-  toCBOR TranslationLogicErrorInput = encode $ Sum TranslationLogicErrorInput 2
-  toCBOR (RdmrPtrPointsToNothing ptr) = encode $ Sum RdmrPtrPointsToNothing 3 !> To ptr
-  toCBOR TranslationLogicErrorDoubleDatum = encode $ Sum LanguageNotSupported 4
-  toCBOR LanguageNotSupported = encode $ Sum LanguageNotSupported 5
-  toCBOR InlineDatumsNotSupported = encode $ Sum InlineDatumsNotSupported 6
-  toCBOR ReferenceScriptsNotSupported = encode $ Sum ReferenceScriptsNotSupported 7
-  toCBOR ReferenceInputsNotSupported = encode $ Sum ReferenceInputsNotSupported 8
-  toCBOR TimeTranslationPastHorizon = encode $ Sum TimeTranslationPastHorizon 9
-
-instance FromCBOR TranslationError where
+instance CC.Crypto crypto => ToCBOR (TxOutSource crypto) where
+  toCBOR = \case
+    TxOutFromInput txIn -> encode $ Sum TxOutFromInput 0 !> To txIn
+    TxOutFromOutput txIx -> encode $ Sum TxOutFromOutput 1 !> To txIx
+
+instance CC.Crypto crypto => FromCBOR (TxOutSource crypto) where
+  fromCBOR = decode (Summands "TxOutSource" dec)
+    where
+      dec 0 = SumD TxOutFromInput <! From
+      dec 1 = SumD TxOutFromOutput <! From
+      dec n = Invalid n
+
+data TranslationError crypto
+  = ByronTxOutInContext !(TxOutSource crypto)
+  | TranslationLogicMissingInput !(TxIn crypto)
+  | RdmrPtrPointsToNothing !RdmrPtr
+  | LanguageNotSupported !Language
+  | InlineDatumsNotSupported !(TxOutSource crypto)
+  | ReferenceScriptsNotSupported !(TxOutSource crypto)
+  | ReferenceInputsNotSupported !(Set (TxIn crypto))
+  | TimeTranslationPastHorizon !Text
```

Comment:
at least `LanguageNotSupported` and `TimeTranslationPastHorizon` are used in the Alonzo era, so we can't change the serialization.

Reply:
I pretty sure this is an incorrect statement. Both of these type constructors where introduced after 1.34.1, which is the latest released version of node.

In 1.34.1:

* There was no costmodel for PlutusV2, so `collectTwoPhaseScriptInputs` returned `NoCostModel` error if anyone where to supply PlutusV2 (whichprobably wasn't possible either): https://github.com/input-output-hk/cardano-ledger/blob/1a9ec4ae9e0b09d54e49b2a40c4ead37edadcce5/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs#L173-L177
In Alonzo `collectTwoPhaseScriptInputs` did not rely on txInfo for validation, because it did not fail at all: https://github.com/input-output-hk/cardano-ledger/blob/1a9ec4ae9e0b09d54e49b2a40c4ead37edadcce5/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L351-L358

* And `TimeTranslationErrorPasHorizon` was not even a thing until 2 weeks ago, because `epochInfoSlotToUTCTime` didn't use to fail until this PR: https://github.com/input-output-hk/cardano-base/pull/277

Reply:
So, I am pretty confident we are free to change this serialization of this type as we wish before the next release, which is the primary reason I went on fixing it in the first place :smile: